Bugfix device map detr model

pedrogengo opened the pull request:
# What does this PR do?

Fixes #26700 #23145


## Who can review?

@LysandreJik @SunMarc

pedrogengo committed:
Fixed replace_batch_norm when on meta device

pedrogengo committed:
lint fix

pedrogengo committed:
Adding coauthor

Co-authored-by: Pi Esposito <piero.skywalker@gmail.com>

SunMarc reviewed (commented):
Hi @pedrogengo thanks for making detr model compatible with `device_map` ! The PR looks good overall. I don't think that we need to add tests for device_map because we already have them. You just need to make sure that you can pass the accelerate [tests](https://huggingface.co/docs/transformers/testing#run-accelerate-tests) for all models. This will trigger the following [tests](https://github.com/huggingface/transformers/blob/fdaef3368b7495f6d3f26739fece0ee370fa7ce6/tests/test_modeling_common.py#L2514C1-L2514C1): `test_disk_offload`, `test_cpu_offload` and  `test_model_parallelism`. Please use 2 GPUs to tests them. If you don't have them, i can run the tests for you!

pedrogengo commented:
Hi @SunMarc :) I don't have 2 GPUs unfortunately. Can you run on your side? In the meantime I can remove the tests I added + fix the style issue

pedrogengo committed:
Removed tests

pedrogengo committed:
Remove unused deps

pedrogengo committed:
Try to fix copy issue

pedrogengo committed:
try fix copy one more time

pedrogengo commented:
@SunMarc addressed your comment on my side. Could you test?
I also noticed a bad pattern maybe on `check_copies.py`. At first, I tried this:

`# Copied from transformers.models.deformable_detr.modeling_deformable_detr.DeformableDetrPreTrainedModel with DeformableDetr->Deta,DeformableDetrConvEncoder->DetaBackboneWithPositionalEncodings`

And it was breaking the CI, because it first replace `DeformableDetr` to `Deta`, and after search for the next pair, which was `DeformableDetrConvEncoder->DetaBackboneWithPositionalEncodings`. However, as we first replaced, it turns that in the code `DeformableDetrConvEncoder` was changed to `DetaConvEncoder`. Maybe we can sort by length and start replacing from the largest string, and also replace with a placeholder in both codes to avoid replacing more than one time. Just something that I see haha I can propose a PR for that, but I don't think it is that much important.

SunMarc commented on tests/models/detr/test_modeling_detr.py:
```diff
@@ -20,7 +20,13 @@
 
 from transformers import DetrConfig, ResNetConfig, is_torch_available, is_vision_available
-from transformers.testing_utils import require_timm, require_torch, require_vision, slow, torch_device
+from transformers.testing_utils import (
+    require_timm,
+    require_torch,
+    require_vision,
+    slow,
+    torch_device,
+)
```
same

SunMarc commented on tests/models/table_transformer/test_modeling_table_transformer.py:
```diff
@@ -22,7 +22,13 @@
 
 from transformers import ResNetConfig, TableTransformerConfig, is_torch_available, is_vision_available
-from transformers.testing_utils import require_timm, require_torch, require_vision, slow, torch_device
+from transformers.testing_utils import (
+    require_timm,
+    require_torch,
+    require_vision,
+    slow,
+    torch_device,
+)
```
same

SunMarc commented on tests/models/conditional_detr/test_modeling_conditional_detr.py:
```diff
@@ -20,7 +20,13 @@
 
 from transformers import ConditionalDetrConfig, ResNetConfig, is_torch_available, is_vision_available
-from transformers.testing_utils import require_timm, require_torch, require_vision, slow, torch_device
+from transformers.testing_utils import (
+    require_timm,
+    require_torch,
+    require_vision,
+    slow,
+    torch_device,
+)
```
As we don't add additional tests, revert the modification.

SunMarc reviewed (approved):
LGTM !  I've tested the accelerate tests on 2GPUs and they passed. Feel free to raise an issue or a PR to fix the pattern on check_copies.

pedrogengo committed:
Reverted import changes

amyeroberts reviewed (approved):
Thanks for fixing @pedrogengo, and thanks @SunMarc for running the accelerate tests!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26849). All of your documentation changes will be reflected on that endpoint.
